statistics/histogram.go

```diff
@@ -858,6 +867,156 @@ func (idx *Index) getRowCount(sc *stmtctx.StatementContext, indexRanges []*range
+			bkt.Repeat = cntPerVal
+		}
+		newHist.Buckets = append(newHist.Buckets, bkt)
+		switch newHist.Tp.EvalType() {
+		case types.ETString, types.ETDecimal, types.ETDatetime, types.ETTimestamp:
+			newHist.scalars = append(newHist.scalars, oldHist.scalars[boundIdx/2])
+		}
+	}
+	return nil
+}
+
+func (idx *Index) newIndexBySelectivity(sc *stmtctx.StatementContext, statsNode *StatsNode) (*Index, error) {
+	var (
+		ranLowEncode, ranHighEncode []byte
+		err                         error
+	)
+	newIndexHist := &Index{Info: idx.Info}
+	newIndexHist.Histogram = *NewHistogram(idx.ID, int64(float64(idx.NDV)*statsNode.Selectivity), 0, 0, types.NewFieldType(mysql.TypeBlob), chunk.InitialCapacity, 0)
+
+	lowBucketIdx, highBucketIdx := 0, 0
+	var totCnt int64 = 0
+
+	// Bucket bound of index is encoded one, so we need to decode it if we want to calculate the fraction accurately.
+	// TODO: enhance its calculation.
+	// Now just remove the bucket that no range fell in.
+	for _, ran := range statsNode.Ranges {
+		lowBucketIdx = highBucketIdx
+		ranLowEncode, ranHighEncode, err = ran.Encode(sc, ranLowEncode, ranHighEncode)
+		if err != nil {
+			return nil, err
+		}
+		for ; highBucketIdx < idx.Len(); highBucketIdx++ {
+			// Encoded value can only go to its next quickly. So ranHighEncode is actually range.HighVal's PrefixNext value.
+			// So the Bound should also go to its PrefixNext.
+			bucketLowerEncoded := idx.Bounds.GetRow(highBucketIdx * 2).GetBytes(0)
+			if bytes.Compare(ranHighEncode, kv.Key(bucketLowerEncoded).PrefixNext()) < 0 {
+				break
+			}
+		}
+		for ; lowBucketIdx < highBucketIdx; lowBucketIdx++ {
+			bucketUpperEncoded := idx.Bounds.GetRow(lowBucketIdx*2 + 1).GetBytes(0)
+			if bytes.Compare(ranLowEncode, bucketUpperEncoded) <= 0 {
+				break
+			}
+		}
+		if lowBucketIdx >= idx.Len() {
+			break
+		}
+		for i := lowBucketIdx; i < highBucketIdx; i++ {
+			newIndexHist.Bounds.AppendRow(idx.Bounds.GetRow(i * 2))
+			newIndexHist.Bounds.AppendRow(idx.Bounds.GetRow(i*2 + 1))
+			totCnt += idx.bucketCount(i)
+			newIndexHist.Buckets = append(newIndexHist.Buckets, Bucket{Repeat: idx.Buckets[i].Repeat, Count: totCnt})
+			newIndexHist.scalars = append(newIndexHist.scalars, idx.scalars[i])
+		}
+	}
+	return newIndexHist, nil
+}
+
+// NewHistCollBySelectivity creates new HistColl by the given statsNodes.
+func (coll *HistColl) NewHistCollBySelectivity(sc *stmtctx.StatementContext, statsNodes []*StatsNode) (*HistColl, error) {
+	newColl := &HistColl{
+		Columns:       make(map[int64]*Column),
+		Indices:       make(map[int64]*Index),
+		Idx2ColumnIDs: coll.Idx2ColumnIDs,
+		ColID2IdxID:   coll.ColID2IdxID,
+		Count:         coll.Count,
+	}
+	for _, node := range statsNodes {
+		if node.Tp == indexType {
+			idxHist, ok := coll.Indices[node.ID]
+			if !ok {
+				continue
+			}
+			newIdxHist, err := idxHist.newIndexBySelectivity(sc, node)
+			if err != nil {
+				log.Warnf("[Histogram-in-plan]: error happened when calculating row count: %v", err)
```

Comment:
should we return `err` and break the histogram construction once we failed to construct the index histogram?

Reply:
Sounds reasonable. How about to print more info in the error message to help us locate the problem once there is an error? For example, we can print the index and range information?

Reply:
Oh, it's good to log it. But IndexInfo currently have no `String` method... What about log its name and the table name?

Reply:
OK, @winoros PTAL this comment.